File: web/playwright-cucumber-typescript/utils/resilient-locator.ts
import { Locator, Page } from '@playwright/test';

/**
 * Self-healing locator that tries multiple selector strategies.
 * Falls through: data-test-id → role → text → CSS
 */
export class ResilientLocator {
  constructor(private page: Page) {}

  async find(
    selectors: { testId?: string; role?: { role: string; name?: string }; text?: string; css?: string },
    options?: { timeout?: number }
  ): Promise<Locator> {
    const timeout = options?.timeout ?? 5000;

    if (selectors.testId) {
      const loc = this.page.locator(`[data-test-id="${selectors.testId}"]`);
      if (await loc.first().isVisible({ timeout }).catch(() => false)) return loc.first();
    }
    if (selectors.role) {
      const loc = this.page.getByRole(selectors.role.role as any, { name: selectors.role.name });
      if (await loc.first().isVisible({ timeout }).catch(() => false)) return loc.first();
    }
    if (selectors.text) {
      const loc = this.page.getByText(selectors.text);
      if (await loc.first().isVisible({ timeout }).catch(() => false)) return loc.first();
    }
    if (selectors.css) {
      const loc = this.page.locator(selectors.css);
      if (await loc.first().isVisible({ timeout }).catch(() => false)) return loc.first();
    }

    // Final fallback — return first non-null for Playwright's auto-wait to handle
    return this.page.locator(
      selectors.testId ? `[data-test-id="${selectors.testId}"]` :
      selectors.css ?? 'body'
    );
  }
}

File: web/playwright-csharp/Utils/ResilientLocator.cs
using Microsoft.Playwright;

namespace Utils;

/// <summary>
/// Self-healing locator — tries data-test-id → role → text → CSS.
/// </summary>
public class ResilientLocator
{
    private readonly IPage _page;
    private readonly float _timeout;

    public ResilientLocator(IPage page, float timeout = 5000)
    {
        _page = page;
        _timeout = timeout;
    }

    public async Task<ILocator> FindAsync(string? testId = null, AriaRole? role = null, string? roleName = null, string? text = null, string? css = null)
    {
        float per = _timeout / 4;

        if (testId != null)
        {
            var loc = _page.Locator($"[data-test-id=\"{testId}\"]").First;
            if (await loc.IsVisibleAsync(new() { Timeout = per }).ContinueWith(t => !t.IsFaulted && t.Result)) return loc;
        }
        if (role != null)
        {
            var loc = _page.GetByRole(role.Value, new() { Name = roleName }).First;
            if (await loc.IsVisibleAsync(new() { Timeout = per }).ContinueWith(t => !t.IsFaulted && t.Result)) return loc;
        }
        if (text != null)
        {
            var loc = _page.GetByText(text).First;
            if (await loc.IsVisibleAsync(new() { Timeout = per }).ContinueWith(t => !t.IsFaulted && t.Result)) return loc;
        }
        if (css != null)
        {
            return _page.Locator(css).First;
        }

        return _page.Locator($"[data-test-id=\"{testId}\"]");
    }
}
